### PR TITLE
fix: restore req body

### PR DESCRIPTION
**Pull request by sigua-cs**
## Description

 
Fix restore original req.Body
## Pull request type
implements https://github.com/ContentSquare/chproxy/issues/336
 

 

Please check the type of change your PR introduces:
- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

### Checklist

- [ ] Linter passes correctly
- [ ] Add tests which fail without the change (if possible)
- [ ] All tests passing
- [ ] Extended the README / documentation, if necessary

## Does this introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

**Review by mga-chka — COMMENTED** [review]
Is there an easy way to add a test in order to assess the before/after change behavior (to ensure we don't have a regression in the future)?

File: proxy.go [hunk 1]
```diff
@@ -207,23 +207,22 @@ func executeWithRetry(
...
 	}
 
 	numRetry := 0
 	for {
-		// update body
+		rp(rw, req)
+
+		// Restore req.Body after it's consumed by 'rp' for potential reuse.
 		req.Body = io.NopCloser(bytes.NewBuffer(body))
 		req.Body.Close()
```
Comment (mga-chka) [inline proxy.go]: why do you have a call to Close() here and there? https://github.com/ContentSquare/chproxy/pull/397/files#diff-8b7f25bac2ae04c2ce2abb628c6aa28103fd113103a62f3e9d232fa995684915R111

I don't know what should be done, but it doesn't look logic to have inconsistencies for the same use case.

Comment (mga-chka) [reply proxy.go]: same comment for https://github.com/ContentSquare/chproxy/pull/397/files#diff-8b7f25bac2ae04c2ce2abb628c6aa28103fd113103a62f3e9d232fa995684915R317

Comment (sigua-cs) [reply proxy.go]: You are right. This one was at the beginning. I'll check if can be removed

**Comment by sigua-cs** [comment]
> Is there an easy way to add a test in order to assess the before/after change behavior (to ensure we don't have a regression in the future)?

I'll try

File: proxy.go (hunk 1 above)
Comment (sigua-cs) [reply proxy.go]: req.Body.Close() appears in few places. Maybe, better to refactor it in a separate PR?

File: proxy_test.go [hunk 2]
```diff
@@ -1012,8 +1012,63 @@ func TestReverseProxy_ServeHTTP2(t *testing.T) {
...
+		resp.Body.Close()
+
+		if !strings.Contains(b, expected) {
+			t.Fatalf("expected response: %q; got: %q", expected, b)
+		}
+
+	})
 }
 
+//func TestRequestBody(t *testing.T) {
```
Comment (mga-chka) [inline proxy_test.go]: Can you remove the commented code?

**Review by mga-chka — APPROVED** [review]

**Review by mga-chka — APPROVED** [review]

**Pull request merged**